### ISSUE TITLE
logging with a configurable in ipython

### ISSUE THREAD

**Issue by pwm1234** [ipython/traitlets#142]
I am trying to use traitlets.Configurable and have found that it works great as a base class. But I am unable to get my configurable to log properly when used from IPython. The application logger works just fine when standalone, but there is no output, not even for error, when I instantiate and use it from ipython. (I just updated to the latest canopy 1.6, so I am fairly up to date with everything, but definitely not bleeding edge.)

**Comment by minrk** [comment]
Can you share code that reproduces the issue?

**Comment by pwm1234** [comment]
The code below works differently in ipython (terminal) and ipython notebook. In crafting this example, I learned a little more about the problem, which I think I could avoid if I better understood how to make an Application subclass that worked properly (and identically) when instantiated from both an ipython terminal and ipython notebook. With a trivial example like this it seems silly. But with a real Application that is responsible for gluing together several complex Configurable objects it is not so obvious (to me at least). Here is a simple example that demonstrates my original problem:

```python
from traitlets.config import Configurable, Application
from traitlets import Unicode, Int, Float, List

import sys

class MyClass(Configurable):
    str = Unicode(u'default string', config=True)
    an_int = Int(1, config=True)
    
    def __init__(self, log, **kwargs):
        super(MyClass, self).__init__(**kwargs)
        self._log = log
        
    def debug(self, msg):
        self._log.debug(msg)
        
    def info(self, msg):
        self._log.info(msg)
                
    def warn(self, msg):
        self._log.warn(msg)
                
    def error(self, msg):
        self._log.error(msg)
        
    def test_log(self):
        self.debug('dbg')
        self.info('info')
        self.warn('warn')
        self.error('error')

class MyApp(Application):
    a_float = Float(1.23, config=True)
    
    classes = List([
        MyClass,
    ])

    def __init__(self, **kwargs):
        super(MyApp, self).__init__(**kwargs)
    
    
def main(argv=None):
    cls = MyClass(MyApp.instance().log)
    cls.test_log()
    
def main2(argv=None):
    cls = MyClass(MyApp().log)
    cls.test_log()

print('calling main2')    
main2(['hello'])
print('works with ipython but not notebook')

print('using logger from host application instance')
MyClass(Application.instance().log).test_log()
print('works with ipython but not notebook')

print('using a.test.logger')
import logging
# could also reload and reconfigure, but do not need too
#reload(logging)
#logging.basicConfig(format='%(asctime)s %(levelname)s:%(message)s', level=logging.DEBUG, datefmt='%I:%M:%S')
mylogger = logging.getLogger('a.test.logger')
mylogger.setLevel(logging.INFO)
MyClass(mylogger).test_log()
print('works with both')


#if __name__ == "__main__":
#   main2(sys.argv)
```

**Comment by pwm1234** [comment]
A better example is the one that ships with traitlets in traitlets/examples/myapp.py. Use the following from both an IPython terminal session and an IPython notebook. The terminal session generates log output while the notebook session does not.

````python
import myapp
import sys
sys.argv = [sys.argv[0]]
sys.argv.append('--MyApp.log_level=DEBUG')
myapp.main()
````

The above code from terminal session generates

````
[MyApp] Config changed:
[MyApp] {'MyApp': {'log_level': u'DEBUG'}}
app.config:
{'MyApp': {'log_level': u'DEBUG'}}
````

and from the notebook session generates, where the `[MyApp]` log output is missing

````
app.config:
{'MyApp': {'log_level': u'DEBUG'}}
````

**Comment by rmorshea** [comment]
I'm not quite sure what's going on with your situation, and I can't imagine it has anything to do with the version of the notebook you're using, but for me, the code you wrote above gives the appropriate output.

**Comment by pwm1234** [comment]
For what it is worth

````python
import IPython
IPython.__version__
````
outputs `'4.0.1'`

**Comment by rmorshea** [comment]
I'm running off master, so `4.1.0-dev`.

**Comment by rkern** [comment]
*Possibly* fixed by #233.
